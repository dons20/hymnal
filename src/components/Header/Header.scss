.page-header {
  width: 100%;
  & .search {
    width: 100%;
    max-width: 30rem;
  }
}

@media only screen and (max-width: 575px) {
  .page-header {
    & .search {
      max-width: 90vw;
    }
  }
  .ant-page-header-heading-title {
    float: none;
    text-align: center;
  }
}
