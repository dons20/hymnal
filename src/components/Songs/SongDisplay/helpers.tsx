import React from 'react';
import { Box, Text } from '@mantine/core';

/**
 * Helper function to render chorus with vocal parts and indicators
 */
export const renderChorusWithParts = (song: Song, isDark: boolean) => {
  if (!song.chorusParts || song.chorusParts.length === 0) {
    // Fallback to regular chorus rendering
    return <Text className="song-text">{song.chorus}</Text>;
  }

  return (
    <Box className="chorus-parts">
      {song.chorusParts.map((part, index) => (
        <Box
          key={index}
          className="chorus-part"
          style={{ position: 'relative', marginBottom: '0.5rem' }}
        >
          {part.part && (
            <Text
              component="span"
              className="part-indicator-regular"
              style={
                {
                  '--part-indicator-color':
                    part.part === 'soprano'
                      ? isDark
                        ? '#74c0fc'
                        : '#1c7ed6' // Blue for soprano
                      : isDark
                        ? '#ffec99'
                        : '#fd7e14', // Orange for baritone
                  '--part-indicator-bg':
                    part.part === 'soprano'
                      ? isDark
                        ? 'rgba(116, 192, 252, 0.15)'
                        : 'rgba(28, 126, 214, 0.08)'
                      : isDark
                        ? 'rgba(255, 236, 153, 0.15)'
                        : 'rgba(253, 126, 20, 0.08)',
                  '--part-indicator-border':
                    part.part === 'soprano'
                      ? isDark
                        ? 'rgba(116, 192, 252, 0.3)'
                        : 'rgba(28, 126, 214, 0.25)'
                      : isDark
                        ? 'rgba(255, 236, 153, 0.3)'
                        : 'rgba(253, 126, 20, 0.25)',
                } as React.CSSProperties
              }
            >
              {part.part === 'soprano' ? 'A' : part.part === 'baritone' ? 'B' : '♪'}
            </Text>
          )}
          <Text
            className="song-text"
            style={
              {
                paddingLeft: part.part ? '0' : '0',
                fontWeight: part.part === 'soprano' ? 500 : 400,
                '--part-text-color':
                  part.part === 'soprano'
                    ? isDark
                      ? '#e9ecef'
                      : '#212529' // Standard text color for soprano
                    : isDark
                      ? '#ced4da'
                      : '#495057', // Slightly muted for baritone
                color: 'var(--part-text-color)',
              } as React.CSSProperties
            }
          >
            {part.text}
          </Text>
        </Box>
      ))}
    </Box>
  );
};

/**
 * Helper function to render presentation slide content (verses or chorus with vocal parts)
 */
export const renderPresentationContent = (content: string | VocalPart[], isDark: boolean) => {
  if (typeof content === 'string') {
    return <Text className="presentation-text">{content}</Text>;
  }

  // Handle VocalPart[] for presentation mode with larger styling
  return (
    <Box className="presentation-content-container">
      {content.map((part, index) => {
        // Check if this is the end of a B part followed by an A part (new pair starting)
        const isEndOfPair =
          part.part === 'baritone' &&
          index < content.length - 1 &&
          content[index + 1]?.part === 'soprano';

        return (
          <React.Fragment key={index}>
            <Box
              className="presentation-part-container"
              style={{
                marginBottom: isEndOfPair ? '2rem' : '1.5rem',
              }}
            >
              {part.part && (
                <Text
                  component="span"
                  className="part-indicator-presentation"
                  style={
                    {
                      '--part-indicator-color':
                        part.part === 'soprano'
                          ? isDark
                            ? '#74c0fc'
                            : '#1c7ed6' // Blue for soprano
                          : isDark
                            ? '#ffec99'
                            : '#fd7e14', // Orange for baritone
                      '--part-indicator-bg':
                        part.part === 'soprano'
                          ? isDark
                            ? 'rgba(116, 192, 252, 0.15)'
                            : 'rgba(28, 126, 214, 0.08)'
                          : isDark
                            ? 'rgba(255, 236, 153, 0.15)'
                            : 'rgba(253, 126, 20, 0.08)',
                      '--part-indicator-border':
                        part.part === 'soprano'
                          ? isDark
                            ? 'rgba(116, 192, 252, 0.3)'
                            : 'rgba(28, 126, 214, 0.25)'
                          : isDark
                            ? 'rgba(255, 236, 153, 0.3)'
                            : 'rgba(253, 126, 20, 0.25)',
                    } as React.CSSProperties
                  }
                >
                  {part.part === 'soprano' ? 'A' : part.part === 'baritone' ? 'B' : '♪'}
                </Text>
              )}
              <Text
                className="presentation-part-text"
                style={
                  {
                    fontWeight: part.part === 'soprano' ? 600 : 400,
                    '--part-text-color':
                      part.part === 'soprano'
                        ? isDark
                          ? '#e9ecef'
                          : '#212529' // Standard text color for soprano
                        : isDark
                          ? '#ced4da'
                          : '#495057', // Slightly muted for baritone
                  } as React.CSSProperties
                }
              >
                {part.text}
              </Text>
            </Box>
            {isEndOfPair && (
              <Box
                className="pair-divider"
                style={
                  {
                    '--divider-color': isDark
                      ? 'rgba(255, 255, 255, 0.2)'
                      : 'rgba(0, 0, 0, 0.15)',
                  } as React.CSSProperties
                }
              />
            )}
          </React.Fragment>
        );
      })}
    </Box>
  );
};
