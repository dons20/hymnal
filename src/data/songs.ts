export const version = "1.1.0";
export const SongsDB = [
	{
		number: 1,
		title: "There Was One Who Was Willing",
		verse: [
			"There was one who was willing to die in my stead,\nThat a soul so unworthy might live,\nAnd the path to the cross He was willing to tread,\nAll the sins of my life to forgive.",
			"He is tender and loving and patient with me,\nWhile He cleanses my heart of its dross,\nBut there's no condemnation, I know I am free,\nFor my sins are all nailed to the cross.",
			"I will cling to my Saviour and never depart,\nI will joyfully journey each day,\nWith a song on my lips and a song in my heart,\nThat my sins have been taken away.",
		],
		chorus: "They are nailed to the Cross, they are nailed to the cross,\nO how much He was willing to bear!\nWith what anguish and loss, Jesus went to the cross!\nBut He carried my sins with Him there.",
		author: "James Montgomery",
	},
	{
		number: 2,
		title: "Redeemed! How I Love To Proclaim It!",
		verse: [
			"Redeemed! How I love to proclaim it!\nRedeemed by the blood of the Lamb;\nRedeemed through His infinite mercy,\nHis child, and forever, I am.",
			"Redeemed! And so happy in Jesus!\nNo language my rapture can tell;\nI know that the light of His presence\nWith me doth continually dwell.",
			"I think of my blessed Redeemer,\nI think of Him all the day long;\nI sing, for I cannot be silent;\nHis love is the theme of my song.",
			"I know I shall see in his beauty\nThe King in whose law I delight,\nWho lovingly guardeth my footsteps,\nAnd giveth me songs in the night.",
			"I know there's a crown that is waiting\nIn yonder bright mansion for me;\nAnd soon, with the spirits made perfect,\nAt home with the Lord I shall be.",
		],
		chorus: "Redeemed, redeemed, redeemed by the blood of the Lamb;\nRedeemed, redeemed, His child and forever, I am.",
		author: "Fanny J. Crosby",
	},
	{
		number: 3,
		title: "O Now I See The Crimson Wave",
		verse: [
			"O now I see the crimson wave, the fountain deep and wide;\nJesus, my Lord, mighty to save, points to His wounded side.",
			"I see the new creation rise, I hear the speaking blood;\nIt speaksâ€”polluted nature dies, sinks 'neath the cleansing flood.",
			"I rise to walk in heaven's own light, above the world and sin;\nWith heart made pure and garments white, and Christ enthroned within.",
			"Amazing grace! 'tis heaven below to feel the blood applied,\nAnd Jesus, only Jesus, know, my Jesus crucified.",
		],
		chorus: "The cleansing stream I see, I see,\nI plunge, and O, it cleanseth me!\nO praise the Lord! It cleanseth me,\nIt cleanseth me, yes, cleanseth me.",
	},
	{
		number: 4,
		title: "Dying With Jesus",
		verse: [
			"Dying with Jesus, by death reckoned mine,\nLiving with Jesus a new life divine,\nLooking to Jesus till glory doth shine,\nMoment by moment O Lord, I am Thine.",
			"Never a trial that He is not there,\nNever a burden that He doth not bear,\nNever a sorrow that He doth not share,\nMoment by moment I'm under His care.",
			"Never a heartache, and never a groan,\nNever a teardrop and never a moan;\nNever a danger but there on the throne,\nMoment by moment He thinks of His own.",
			"Never a weakness that He doth not feel,\nNever a sickness that He cannot heal;\nMoment by moment, in woe or in weal,\nJesus, my Saviour, abides with me still.",
		],
		chorus: "Moment by moment I'm kept in His love;\nMoment by moment I've life from above;\nLooking to Jesus till glory doth shine;\nMoment by moment, O Lord, I am Thine.",
		author: "D. W. Whittle",
	},
	{
		number: 5,
		title: "Come, Every Soul By Sin Oppressed",
		verse: [
			"Come, every soul by sin oppressed, there's mercy with the Lord,\nAnd He will surely give you rest, by trusting in His word.",
			"For Jesus shed His precious blood rich blessings to bestow;\nPlunge now into the crimson flood that washes white as snow.",
			"Yes, Jesus is the truth, the way, that leads you into rest;\nBelieve in Him without delay, and you are fully blest.",
			"Come, then, and join this holy band, and on to glory go,\nTo dwell in that celestial land, where joys immortal flow.",
		],
		chorus: "Only trust Him, only trust Him, only trust Him now;\nHe will save you, He will save you, He will save you now.",
		author: "J. H. Stockton",
	},
];
