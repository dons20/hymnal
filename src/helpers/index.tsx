/* eslint-disable no-console */
import React from "react";
import axios from "axios";
import hymns from "static/hymns.json";
import setDebug from "utils/logger";

interface SongPropsA {
	number: number;
	title: string;
	verse: string[];
	chorus: string;
	author?: string;
}
interface SongPropsB {
	number: number;
	title: string;
	verse: string[];
	chorus: string;
	author: string;
}

export type SongProps = SongPropsA | SongPropsB;

export type SongsDB = { songs: SongProps[]; version: string };

setDebug();
const { debug } = window;

/**
 * Loads songs from JSON and stores them locally
 */
export async function loadNewSongs(fetchedSongs: SongProps[]) {
	const localForage = await import("localforage");
	const songs = localForage.createInstance({ storeName: "items" });
	const localVersion = localForage.createInstance({ storeName: "version" });
	const favourites = localForage.createInstance({ storeName: "favourites" });

	try {
		await Promise.all([
			fetchedSongs.forEach((song, i) => songs.setItem(`${i}`, { ...song }).catch(e => debug.info(e))),
			localVersion.setItem("value", localVersion).catch(e => debug.info(e)),
			favourites.clear(),
		]);
	} catch (err) {
		if (err instanceof Error) debug.info(err.message);
	}
}

/**
 * Checks if songs have already been stored
 */
export async function checkDB() {
	const dbName = "Songs";
	const localForage = await import("localforage");
	localForage.config({
		name: dbName,
		description: "Stores the songs db and its version number",
	});

	let songsDB: SongsDB;
	try {
		const fetchedData = process.env.NODE_ENV === "production" ? await axios.get<SongsDB>(process.env.HYMNS_URL || "https://f002.backblazeb2.com/file/hymnal/hymns.json") : { data: hymns };
		if (!fetchedData.data) return;
		songsDB = fetchedData.data;

		debug.log(`%cChecking if songs exist already`, "color: #3182ce; font-size: medium;");
		const songs = localForage.createInstance({ storeName: "items" });
		const songsLength = await songs.length();
		if (songsLength < songsDB.songs.length) {
			debug.info("Items out of sync with latest items");
			loadNewSongs(songsDB.songs);
		}

		debug.log(`%cChecking for updates`, "color: #3182ce; font-size: medium;");
		const localVersion = localForage.createInstance({ storeName: "version" });
		if (!localVersion) throw new Error("No version stored");
		const versionNumber = (await localVersion.getItem("value")) as string;
		if (songsDB.version !== versionNumber) {
			debug.info("Version mismatch, sync necessary");
			loadNewSongs(songsDB.songs);
		}

		debug.log(`%cSongs found! Attempting to load...`, "color: #3182ce; font-size: medium;");
	} catch (e) {
		if (e instanceof Error) debug.error(e.message);
	}
}

export function useQuery(search: string) {
	return new URLSearchParams(search);
}

/**
 * A helper to create a Context and Provider with no upfront default value, and
 * without having to check for undefined all the time.
 */
export function createCtx<A extends {} | null>() {
	const ctx = React.createContext<A | undefined>(undefined);
	function useCtx() {
		const c = React.useContext(ctx);
		if (c === undefined) throw new Error("useCtx must be inside a Provider with a value");
		return c;
	}
	return [useCtx, ctx.Provider, ctx.Consumer] as const; // 'as const' makes TypeScript infer a tuple
}

export async function updateFavesDB(indexes: number[]) {
	const name = "Songs";
	const storeName = "Favourites";
	const localForage = await import("localforage");

	localForage.config({
		name,
		storeName,
		description: "Your favourite songs",
	});

	indexes.forEach(async item => {
		await localForage.setItem(`${item}`, item);
	});
}
