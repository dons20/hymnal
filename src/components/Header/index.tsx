import React, { lazy, useState } from "react";
import { useColorMode, useMediaQuery, useDisclosure, useColorModeValue, Box } from "@chakra-ui/react";
import { FaSearch, FaSun, FaMoon, FaHome, FaBars } from "react-icons/fa";
import { useDebouncedCallback } from "use-debounce";
import withSuspense from "helpers/withSuspense";
import { useMainContext } from "utils/context";
import { useNavigate } from "react-router-dom";
import Fuse, { FuseResult } from "fuse.js";
import "./Header.scss";

import type {
	Icon as IconType,
	Grid as GridType,
	Text as TextType,
	Fade as FadeType,
	Input as InputType,
	Modal as ModalType,
	Portal as PortalType,
	VStack as VStackType,
	Heading as HeadingType,
	ModalBody as ModalBodyType,
	IconButton as IconButtonType,
	InputGroup as InputGroupType,
	ModalHeader as ModalHeaderType,
	CloseButton as CloseButtonType,
	ModalOverlay as ModalOverlayType,
	ModalContent as ModalContentType,
	ModalCloseButton as ModalCloseButtonType,
	InputRightElement as InputRightElementType,
} from "@chakra-ui/react/dist";

/* Lazy Base Imports */
const CloseButtonImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.CloseButton })));
const FadeImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Fade })));
const GridImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Grid })));
const HeadingImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Heading })));
const InputImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Input })));
const IconImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Icon })));
const IconButtonImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.IconButton })));
const InputGroupImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.InputGroup })));
const InputRightElementImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.InputRightElement })));
const ModalImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Modal })));
const ModalOverlayImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.ModalOverlay })));
const ModalContentImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.ModalContent })));
const ModalHeaderImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.ModalHeader })));
const ModalCloseButtonImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.ModalCloseButton })));
const ModalBodyImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.ModalBody })));
const PortalImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Portal })));
const TextImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Text })));
const VStackImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.VStack })));
const ButtonImport = lazy(() => import("components/Button"));

/* With Suspense Wrapper */
const Fade = withSuspense<typeof FadeType>(FadeImport);
const Grid = withSuspense<typeof GridType, null>(GridImport, null);
const Heading = withSuspense<typeof HeadingType, null>(HeadingImport, null);
const Input = withSuspense<typeof InputType, null>(InputImport, null);
const Icon = withSuspense<typeof IconType>(IconImport);
const CloseButton = withSuspense<typeof CloseButtonType>(CloseButtonImport);
const IconButton = withSuspense<typeof IconButtonType>(IconButtonImport);
const InputGroup = withSuspense<typeof InputGroupType, null>(InputGroupImport, null);
const InputRightElement = withSuspense<typeof InputRightElementType, null>(InputRightElementImport, null);
const Modal = withSuspense<typeof ModalType, null>(ModalImport, null);
const ModalOverlay = withSuspense<typeof ModalOverlayType, null>(ModalOverlayImport, null);
const ModalContent = withSuspense<typeof ModalContentType, null>(ModalContentImport, null);
const ModalHeader = withSuspense<typeof ModalHeaderType, null>(ModalHeaderImport, null);
const ModalCloseButton = withSuspense<typeof ModalCloseButtonType, null>(ModalCloseButtonImport, null);
const ModalBody = withSuspense<typeof ModalBodyType, null>(ModalBodyImport, null);
const Portal = withSuspense<typeof PortalType, null>(PortalImport, null);
const Text = withSuspense<typeof TextType>(TextImport);
const VStack = withSuspense<typeof VStackType, null>(VStackImport, null);
const Button = withSuspense<typeof ButtonImport, null>(ButtonImport, null);

function Header() {
	const navigate = useNavigate();
	const { songs } = useMainContext();
	const { colorMode, toggleColorMode } = useColorMode();
	const { isOpen, onOpen, onClose } = useDisclosure();
	const { isOpen: isModalOpen, onOpen: onModalOpen, onClose: onModalClose } = useDisclosure();
	const fuse = new Fuse(songs!, { keys: ["number", "title"], minMatchCharLength: 2, threshold: 0.4 });
	const [query, setQuery] = useState("");
	const [mobileQuery, setMobileQuery] = useState("");
	const [queryResults, setQueryResults] = useState<FuseResult<Song>[]>([]);
	const [showMobileMenu] = useMediaQuery("(max-width: 550px)");
	const resultsBG = useColorModeValue("white", "gray.800");
	const headerBG = useColorModeValue("gray.100", "gray.800");
	const searchBG = useColorModeValue("gray.50", "gray.700");

	/** Will navigate one level up in the application */
	const GoHome = () => navigate("/");

	const searchSongs = (_?: React.ChangeEvent<any>, mobile?: boolean) => {
		const shouldSearchMobile = mobile && mobileQuery.length > 0;
		const shouldSearchDesktop = !mobile && query.length > 0;
		if (shouldSearchMobile || shouldSearchDesktop) {
			onClose();
			onModalClose();
			setQuery("");
			setMobileQuery("");
			setQueryResults([]);
			navigate(`/search?query=${shouldSearchMobile ? mobileQuery : query}`);
		}
	};

	const submitQuery = (e: React.FormEvent<HTMLDivElement>) => {
		e.preventDefault();
		if (query.length > 0) searchSongs();
	};
	const submitMobileQuery = (e: React.FormEvent<HTMLDivElement>) => {
		e.preventDefault();
		if (mobileQuery.length > 0) searchSongs(e, true);
	};

	const mobileSearchQueryChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		const searchValue = e.target.value;
		setMobileQuery(searchValue);
	};

	const handleSearch = useDebouncedCallback((value: string) => {
		if (value.length === 0) {
			onClose();
			return;
		}
		const result = fuse.search(value, { limit: 6 });
		setQueryResults(result.slice(0, 6));
		if (result.length > 0) onOpen();
	}, 500);

	const searchQueryChange = (value: string) => {
		setQuery(value);
		handleSearch(value);
	};

	const gotoSong = (index: number) => {
		onClose();
		setQuery("");
		navigate(`/songs/${index}`);
	};

	return (
		<Box className="page-header" p={4} bg={headerBG}>
			<Grid templateColumns="auto 1fr" alignItems="center" gap={10} justifyContent="space-between" px={5}>
				<Heading size="md" onClick={GoHome} cursor="pointer" display="flex" alignItems="center" width="auto">
					Hymns for All Times
					<Icon as={FaHome} size={20} ml={3} />
				</Heading>
				{showMobileMenu ? (
					<IconButton
						size="xs"
						icon={<FaBars />}
						aria-label="Open mobile menu"
						justifySelf="flex-end"
						onClick={onModalOpen}
						data-testid="mobileMenuTrigger"
					/>
				) : (
					<Grid templateColumns="minmax(auto, 300px) auto" gap={2} justifyContent="flex-end">
						<InputGroup size="md" as="form" onSubmit={submitQuery} role="search">
							<Input
								value={query}
								onChange={(e: React.ChangeEvent<HTMLInputElement>) => searchQueryChange(e.target.value)}
								type="search"
								placeholder="Search songs..."
								pr="4.5rem"
								bg={searchBG}
								data-testid="desktopSearch"
							/>
							<InputRightElement>
								<IconButton
									size="sm"
									h="1.75rem"
									icon={<FaSearch />}
									aria-label="Search Song Database"
									onClick={searchSongs}
								/>
							</InputRightElement>
						</InputGroup>

						<IconButton
							size="md"
							w="1.5rem"
							onClick={toggleColorMode}
							icon={colorMode === "light" ? <FaMoon /> : <FaSun />}
							aria-label="Toggle Color Mode"
						/>
					</Grid>
				)}
			</Grid>
			<Fade in={isOpen}>
				<Portal>
					<Box
						bg={resultsBG}
						pos="absolute"
						zIndex={105}
						top="60px"
						right="65"
						w={350}
						px={5}
						pb={5}
						hidden={!isOpen || showMobileMenu || undefined}
						borderRadius="md"
						shadow="md"
						data-testid="searchResultsBox"
					>
						<CloseButton size="md" onClick={onClose} pos="absolute" top="1" right="5" />
						<Grid gap={5} mt={10} w="100%" data-testid="searchItemsWrapper">
							{queryResults.map(result => (
								<Button
									bg="blue.500"
									onClick={() => gotoSong(result.item.number)}
									overflow="hidden"
									key={result.item.number}
									role="button"
								>
									<Grid templateColumns="auto 1fr" gap="3" w="100%" justifyItems="left">
										<Text size="sm" color="white">
											{result.item.number}
										</Text>
										<Text size="sm" color="white">
											{result.item.title}
										</Text>
									</Grid>
								</Button>
							))}
						</Grid>
					</Box>
				</Portal>
			</Fade>

			<Modal isOpen={isModalOpen} onClose={onModalClose}>
				<ModalOverlay />
				<ModalContent pb={10}>
					<ModalHeader>Menu</ModalHeader>
					<ModalCloseButton />
					<ModalBody>
						<VStack spacing="4">
							<InputGroup size="md" as="form" onSubmit={submitMobileQuery}>
								<Input
									type="search"
									value={mobileQuery}
									onChange={mobileSearchQueryChange}
									placeholder="Search songs..."
									pr="4.5rem"
									bg={searchBG}
									data-testid="mobileSearch"
								/>
								<InputRightElement>
									<IconButton
										size="sm"
										h="1.75rem"
										icon={<FaSearch />}
										aria-label="Search Song Database"
										onClick={(e: React.MouseEvent<HTMLButtonElement>) => searchSongs(e, true)}
									/>
								</InputRightElement>
							</InputGroup>

							<Button
								leftIcon={colorMode === "light" ? <FaMoon /> : <FaSun />}
								onClick={toggleColorMode}
								w="100%"
							>
								Toggle {colorMode === "light" ? "Dark" : "Light"} Mode
							</Button>
						</VStack>
					</ModalBody>
				</ModalContent>
			</Modal>
		</Box>
	);
}

export default Header;
