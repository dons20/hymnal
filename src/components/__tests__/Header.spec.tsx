import userEvent from "@testing-library/user-event";
import { render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter, BrowserRouter } from "react-router-dom";
import { Providers, resizeWindow } from "helpers/tests";
import { ColorModeScript } from "@chakra-ui/system";
import { ChakraProvider } from "@chakra-ui/react";
import MainContext from "components/Providers";
import { createMemoryHistory } from "history";
import Header from "components/Header";
import { SongsDB } from "data/songs";

const songs = SongsDB;
const user = userEvent.setup();

beforeEach(() => {
    window.matchMedia = jest.fn().mockImplementation(() => ({
        matches: false,
        media: "",
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
    }));
});

describe("#Header", () => {
	it("should render correctly", async () => {
		const { asFragment } = render(
			<BrowserRouter>
				<Providers value={{ songs }}>
					<Header />
				</Providers>
			</BrowserRouter>
		);
		const title = await screen.findByText(/Hymns for All Times/);
		expect(title).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});
	it("should show the search results when typing and hide them when query is erased", async () => {
		render(
			<BrowserRouter>
				<Providers value={{ songs }}>
					<MainContext>
						<Header />
					</MainContext>
				</Providers>
			</BrowserRouter>
		);

		const searchWrapper: HTMLFormElement = await screen.findByRole("search");
		expect(searchWrapper).toBeInTheDocument();

		const searchField: HTMLInputElement = await screen.findByTestId("desktopSearch");
		await user.type(searchField, "test");

		const resultsList = await screen.findByTestId("searchItemsWrapper");

		expect(searchField.value).toBe("test");
		await waitFor(() => expect(resultsList).not.toBeEmptyDOMElement());

		searchField.setSelectionRange(0, 4);
		await user.type(searchField, "{backspace}");

		const searchResultsBox = await screen.findByTestId("searchResultsBox");
		expect(searchResultsBox).toHaveProperty("hidden");
	});
	it("should send the user to a song's page directly when they click a result", async () => {
		render(
			<BrowserRouter>
				<Providers value={{ songs }}>
					<MainContext>
						<Header />
					</MainContext>
				</Providers>
			</BrowserRouter>
		);

		const searchWrapper: HTMLFormElement = await screen.findByRole("search");
		expect(searchWrapper).toBeInTheDocument();

		const searchField: HTMLInputElement = screen.getByTestId("desktopSearch");
		await user.type(searchField, "test");

		const resultsList = screen.getByTestId("searchItemsWrapper");

		await waitFor(() => expect(resultsList.childElementCount).toBeGreaterThan(0));
		const result = await screen.findByText("Lovest Thou Me");
		await user.click(result);

		await waitFor(() => {
			expect(window.location.href).toContain("/songs/399");
		});
	});
	it("should navigate to the search results page when a search is submitted", async () => {
		render(
			<BrowserRouter>
				<Providers value={{ songs }}>
					<Header />
				</Providers>
			</BrowserRouter>
		);

		const searchField = screen.getByTestId("desktopSearch");
		await user.type(searchField, "test");

		await user.type(searchField, "[Enter]");
		await waitFor(() => {
			expect(window.location.href).toContain("/search?query=test");
		});
	});
	it("should toggle night mode when the user clicks the night mode toggle button", async () => {
		render(
			<ChakraProvider>
				<ColorModeScript initialColorMode="light" />
				<MemoryRouter initialEntries={["/home"]}>
					<Providers value={{ songs }}>
						<Header />
					</Providers>
				</MemoryRouter>
			</ChakraProvider>
		);
		const colorModeToggle = await screen.findByLabelText("Toggle Color Mode");

		await waitFor(() => expect(document.body).toHaveClass("chakra-ui-light"));

		await user.click(colorModeToggle);
		await waitFor(() => expect(document.body).toHaveClass("chakra-ui-dark"));
	});
	it("should navigate the user home when they click the main logo", async () => {
		const history = createMemoryHistory();
		render(
			<MemoryRouter>
				<Providers value={{ songs }}>
					<Header />
				</Providers>
			</MemoryRouter>
		);

		const logo = await screen.findByRole("heading", { name: /Hymns For All Times/i });

		await user.click(logo);
		expect(history.location.pathname).toBe("/");
	});
	it("should show a modal popup with a search bar and toggle button on smaller screens", async () => {
		window.matchMedia = jest.fn().mockImplementation(query => ({
			matches: query === "(max-width: 550px)",
			media: "",
			onchange: null,
			addListener: jest.fn(),
			removeListener: jest.fn(),
		}));

		render(
			<BrowserRouter>
				<Providers value={{ songs }}>
					<Header />
				</Providers>
			</BrowserRouter>
		);

		resizeWindow(450, 1000);
		const mobileMenuToggle = await screen.findByTestId("mobileMenuTrigger");
		await user.click(mobileMenuToggle);

		const searchField = await screen.findByTestId("mobileSearch");
		const searchButton = await screen.findByLabelText("Search Song Database");

		await user.type(searchField, "people");
		await user.click(searchButton);

		await waitFor(() => {
			expect(window.location.href).toContain("/search?query=people");
		});
	});
	it.todo("should navigate the user to the search results page when submitting the mobile search field");
	it.todo("should toggle night mode when the user clicks the mobile toggle button");
});
