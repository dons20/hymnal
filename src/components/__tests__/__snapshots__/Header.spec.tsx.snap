// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Header should render correctly 1`] = `
<DocumentFragment>
  <div
    class="page-header css-ln7hi9"
  >
    <div
      class="css-1fianhz"
    >
      <h2
        class="chakra-heading css-1dxtpo3"
      >
        Hymns for All Times
        <svg
          class="chakra-icon css-11gi69r"
          fill="currentColor"
          focusable="false"
          height="20"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 576 512"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
          />
        </svg>
      </h2>
      <div
        class="css-1tlocfr"
      >
        <form
          class="chakra-input__group css-bx0blc"
          data-group="true"
          role="search"
        >
          <input
            class="chakra-input css-mlsdxy"
            data-testid="desktopSearch"
            placeholder="Search songs..."
            type="search"
            value=""
          />
          <div
            class="chakra-input__right-element css-dp9r5f"
          >
            <button
              aria-label="Search Song Database"
              class="chakra-button css-17ywzgp"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                />
              </svg>
            </button>
          </div>
        </form>
        <button
          aria-label="Toggle Color Mode"
          class="chakra-button css-udgpmr"
          type="button"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 160c-52.9 0-96 43.1-96 96s43.1 96 96 96 96-43.1 96-96-43.1-96-96-96zm246.4 80.5l-94.7-47.3 33.5-100.4c4.5-13.6-8.4-26.5-21.9-21.9l-100.4 33.5-47.4-94.8c-6.4-12.8-24.6-12.8-31 0l-47.3 94.7L92.7 70.8c-13.6-4.5-26.5 8.4-21.9 21.9l33.5 100.4-94.7 47.4c-12.8 6.4-12.8 24.6 0 31l94.7 47.3-33.5 100.5c-4.5 13.6 8.4 26.5 21.9 21.9l100.4-33.5 47.3 94.7c6.4 12.8 24.6 12.8 31 0l47.3-94.7 100.4 33.5c13.6 4.5 26.5-8.4 21.9-21.9l-33.5-100.4 94.7-47.3c13-6.5 13-24.7.2-31.1zm-155.9 106c-49.9 49.9-131.1 49.9-181 0-49.9-49.9-49.9-131.1 0-181 49.9-49.9 131.1-49.9 181 0 49.9 49.9 49.9 131.1 0 181z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="chakra-fade"
      style="opacity: 0;"
    />
  </div>
</DocumentFragment>
`;
